java/ql/src/Telemetry/AutomodelApplicationModeCharacteristics.qll

```diff
@@ -0,0 +1,444 @@
+/**
+ * For internal use only.
+ */
+
+private import java
+private import semmle.code.Location as Location
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.dataflow.TaintTracking
+private import semmle.code.java.security.PathCreation
+private import semmle.code.java.dataflow.ExternalFlow as ExternalFlow
+private import semmle.code.java.dataflow.internal.FlowSummaryImpl as FlowSummaryImpl
+private import semmle.code.java.security.ExternalAPIs as ExternalAPIs
+private import semmle.code.java.Expr as Expr
+private import semmle.code.java.security.QueryInjection
+private import semmle.code.java.security.RequestForgery
+private import semmle.code.java.dataflow.internal.ModelExclusions as ModelExclusions
+private import AutomodelJavaUtil as AutomodelJavaUtil
+private import semmle.code.java.security.PathSanitizer as PathSanitizer
+private import AutomodelSharedGetCallable as AutomodelSharedGetCallable
+import AutomodelSharedCharacteristics as SharedCharacteristics
+import AutomodelEndpointTypes as AutomodelEndpointTypes
+
+newtype JavaRelatedLocationType = CallContext()
+
+/**
+ * A class representing nodes that are arguments to calls.
+ */
+private class ArgumentNode extends DataFlow::Node {
+  ArgumentNode() { this.asExpr() = [any(Call c).getAnArgument(), any(Call c).getQualifier()] }
```

Comment:
I've been thinking about this. This isn't taking into account variadic arguments, which means that, for the following:

```java
public void test(String a, String... b) {}
public void test2() { test("a", "b", "c"); }
```

The arguments of `test` will have the following positions and generate the respective MaD "input" strings, if I understand the current CodeQL implementation correctly:


| Expr | idx | input |
| ---- | --- | ----- |
| test(...) | -1 | Argument[-1] |
| "a" | 0 | Argument[0] |
| "b" | 1 | Argument[1] |
| "c" | 2 | Argument[2] |

However, the signature of `test` is `test(String,String[])`, which means that `Argument[2]` won't match anything. In this case, both "b" and "c" should generate `Argument[1]` since they are part of the same variadic argument.

Note that the kind of `DataFlow::Node` that handles this isn't an `ExprNode` but rather an `ImplicitVarargsArray`. That's why using `Node::asExpr` won't work well in those cases.

Happy to handle this in a different PR if you want this merged ASAP, as long as you're aware of this shortcoming in the current implementation.

Reply:
Thanks @atorralb! I've added your comment to [this list](https://github.com/github/CodeML/issues/117) so that we don't forget about it but my understanding is that it shouldn't prevent us from merging this for now.